### DISCUSSION TITLE
Blank line among function completions with llama installed

### DISCUSSION THREAD

**Discussion by meedstrom** [minad/vertico#481] [Q&A] [answered]
Hallo! This is a rather special-case issue, one that seems to be shared with Helm, but not with fido-vertical-mode.

Can reproduce in emacs -Q as follows, but you can probably also see it just by installing [llama](https://git.sr.ht/~tarsius/llama) on your current setup.

```
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(package-initialize)
(package-install 'llama)
(package-install 'vertico)
(vertico-mode)
```

Ok, so you installed llama, now do C-h f (describe-function).

There is a blank line amongst the completions.  It arises from llama's `(defmacro ## ...)`.  Not sure how to fix that.

Ideally there would be a completion ##, but I understand if that's out of scope for vertico. But I guess you don't want a blank line that's always at the top of the candidates.

A difference with Helm is Helm doesn't necessarily put it at the top of the list.

**Comment by minad** [comment, marked as answer]
Vertico moves exactly matching candidates to the front:

https://github.com/minad/vertico/blob/b413071021a717bc25f730d2f94ce0740a9b8e34/vertico.el#L344

If we have an empty candidate, it will appear at the top, if the prompt is empty. I am not sure if it makes sense to add a special case only for this.

**Comment by meedstrom** [comment]
But it is technically not an exact match, so I would regard it as a bug. Because llama does not introduce a 0-length macro. The macro is `##`, and I have not typed in "##".

**Reply by minad** [reply]
`##` is the notation for the empty symbol. You can try `(intern "") => '##` and `(symbol-name '##) => ""`.

**Comment by meedstrom** [comment]
Fair. So the issue is really with Llama, or with emacs-devel.

On that note, I wonder if you know of some trick that Llama could do, in a minor mode or so, to exclude the empty string from all completion candidates?

**Reply by minad** [reply]
From my pov, there is no issue. There is an empty symbol and this is what you see. There is no way to fix this on a lower level since the equality `(equal (symbol-name (intern x)) x)` must hold. This means that `(symbol-name '##)` cannot return a string like `"##"` as you might have expected. Maybe llama could use another symbol, but as far as I know it specifically uses `##` for more concise notation.

**Comment by meedstrom** [comment]
Even if there is no technical issue, you can still have an usability issue.

Anyway I'm not looking at Vertico as the problem domain, now I think it's actually with `help--symbol-completion-table`, called by describe-function... or actually `completing-read` itself.  Think about it, why should a completing-read ever include an empty string among the candidates, knowing that you're supposed to be able to get a default result DEF when you enter the empty string?  There's no way to distinguish what the user intended.

Actually...! Completing-read already behaves correctly. If you turn off vertico-mode, `describe-function` does indeed not take the empty string to match the ## macro,  it just says  "You didn't specify a function symbol". 

Vertico and Helm actually do something strange, in that you can choose an empty string that's not the null input, but base completing-read offers no such possibility. So it looks to me like an unintended extension to completing-read behavior.

**Reply by minad** [reply]
> Actually...! Completing-read already behaves correctly. If you turn off vertico-mode, describe-function does indeed not take the empty string to match the ## macro, it just says "You didn't specify a function symbol".

Default completion behaves the same as Vertico. If you exit with the empty candidate in Vertico, you will also see the error "You didn't specify a function symbol". The reason for this behavior in `completing-read` in general is that the empty string is also treated as the so-called "null completion", which means that the user exited with empty input.

>  Think about it, why should a completing-read ever include an empty string among the candidates, knowing that you're supposed to be able to get a default result DEF when you enter the empty string? There's no way to distinguish what the user intended.

Yes, it better should not, due to the indistinguishability of the "null completion".

> Vertico and Helm actually do something strange, in that you can choose an empty string that's not the null input, but base completing-read offers no such possibility. So it looks to me like an unintended extension to completing-read behavior.

No, there is nothing strange in Vertico, Helm or default completion. If you press `C-h f TAB` (and wait for a while) in default completion you will also see a list of candidates in the completion buffer, including the empty candidate. The same in Vertico or Helm where you see the empty candidate, just more directly.

I think the correct fix of this problem would be to simply disallow the empty symbol at all, since it may lead to edge cases at multiple places. But this is certainly a too radical proposal and it would also break llama. Clearly, `describe-function` doesn't expect function symbols to be empty, leading to the "You didn't specify a function symbol" error message.

**Reply by meedstrom** [reply]
Thanks for the example with `C-h f TAB`! So I'm prepared to just consider llama a dead end (because any package depending on it force all users to see this blank line).

The only real patch that Llama could do is advise `help--symbol-completion-table`.

**Reply by meedstrom** [reply]
Paging @tarsius in case he wants to think about that.  I think the defmacro would need to stop being autoloaded too.

**Reply by tarsius** [reply]
A possible workaround would be to `unintern` the symbol during completion. I've pushed a draft to the `work` branch. Unfortunate, but not horrible.

**Reply by wyuenho** [reply]
This macro has introduced the same problem to helpful.el as well. Just dropping a PR here in case any helpful.el users wonders what that empty string at the top of every `helpful-macro` and `helpful-callable` command invocation is.

https://github.com/Wilfred/helpful/pull/364
